File: src-tauri/src/commands/deals.rs
// ── Deal / Pipeline Commands ────────────────────────────────────────────────

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tauri::State;
use sqlx::SqlitePool;
use crate::db::error::AppDbError;
use crate::db::tables::crm::deals::{
    self, CreateDealRequest, Deal, DealStage, Pipeline,
};
use crate::db::tables::crm::scoring;
use crate::error::SerializedError;

type CmdResult<T> = Result<T, SerializedError>;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateDealCmd {
    pub company_id: String,
    pub contact_id: Option<String>,
    pub pipeline_id: String,
    pub stage_id: String,
    pub title: String,
    pub amount_minor: i64,
    pub currency: Option<String>,
    pub expected_close_at: Option<i64>,
    pub notes: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateDealCmd {
    pub company_id: String,
    pub id: String,
    #[serde(default)]
    pub fields: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatePipelineCmd {
    pub company_id: String,
    pub name: String,
    pub is_default: Option<bool>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateStageCmd {
    pub pipeline_id: String,
    pub name: String,
    pub position: i64,
    pub probability: i64,
    pub color: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MoveDealStageCmd {
    pub id: String,
    pub stage_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RecomputeScoresCmd {
    pub company_id: String,
}

#[tauri::command]
pub async fn db_create_deal(pool: State<'_, SqlitePool>, cmd: CreateDealCmd) -> CmdResult<Deal> {
    let req = CreateDealRequest {
        company_id: cmd.company_id,
        contact_id: cmd.contact_id,
        pipeline_id: cmd.pipeline_id,
        stage_id: cmd.stage_id,
        title: cmd.title,
        amount_minor: cmd.amount_minor,
        currency: cmd.currency,
        expected_close_at: cmd.expected_close_at,
        notes: cmd.notes,
    };
    deals::create_deal(&pool, req).await.map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_update_deal(pool: State<'_, SqlitePool>, cmd: UpdateDealCmd) -> CmdResult<Deal> {
    deals::update_deal(&pool, &cmd.id, cmd.fields)
        .await
        .map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_delete_deal(pool: State<'_, SqlitePool>, id: String) -> CmdResult<()> {
    deals::delete_deal(&pool, &id).await.map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_get_deal(pool: State<'_, SqlitePool>, id: String) -> CmdResult<Deal> {
    deals::get_deal(&pool, &id)
        .await
        .map_err(|e| SerializedError::from(e))?
        .ok_or_else(|| SerializedError::from(AppDbError::NotFound(format!("deal {id}"))))
}

#[tauri::command]
pub async fn db_list_deals(
    pool: State<'_, SqlitePool>,
    company_id: String,
    pipeline_id: Option<String>,
    stage_id: Option<String>,
    status: Option<String>,
) -> CmdResult<Vec<Deal>> {
    deals::list_deals(&pool, &company_id, pipeline_id.as_deref(), stage_id.as_deref(), status.as_deref())
        .await
        .map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_move_deal_stage(
    pool: State<'_, SqlitePool>,
    cmd: MoveDealStageCmd,
) -> CmdResult<Deal> {
    deals::move_deal_stage(&pool, &cmd.id, &cmd.stage_id)
        .await
        .map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_create_pipeline(
    pool: State<'_, SqlitePool>,
    cmd: CreatePipelineCmd,
) -> CmdResult<Pipeline> {
    deals::create_pipeline(&pool, &cmd.company_id, &cmd.name, cmd.is_default.unwrap_or(false))
        .await
        .map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_list_pipelines(
    pool: State<'_, SqlitePool>,
    company_id: String,
) -> CmdResult<Vec<Pipeline>> {
    deals::list_pipelines(&pool, &company_id).await.map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_create_deal_stage(
    pool: State<'_, SqlitePool>,
    cmd: CreateStageCmd,
) -> CmdResult<DealStage> {
    deals::create_stage(
        &pool,
        &cmd.pipeline_id,
        &cmd.name,
        cmd.position,
        cmd.probability,
        &cmd.color.unwrap_or_else(|| "#0b57d0".to_string()),
    )
    .await
    .map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_list_deal_stages(
    pool: State<'_, SqlitePool>,
    pipeline_id: String,
) -> CmdResult<Vec<DealStage>> {
    deals::list_stages(&pool, &pipeline_id).await.map_err(|e| SerializedError::from(e))
}

#[tauri::command]
pub async fn db_recompute_scores(
    pool: State<'_, SqlitePool>,
    cmd: RecomputeScoresCmd,
) -> CmdResult<usize> {
    scoring::recompute_scores(&pool, &cmd.company_id)
        .await
        .map_err(|e| SerializedError::from(e))
}

File: src/shared/services/db/invoke/deals.ts
// ── Deal / Pipeline wrappers ──────────────────────────────────────────────

import { invokeCommand } from './command';

import type {
  Deal,
  DealStage,
  Pipeline,
  CreateDealInput,
  MoveDealStageInput,
  RecomputeScoresInput,
} from '../schema';

// Pipelines

export async function listPipelines(companyId: string): Promise<Pipeline[]> {
  return invokeCommand<Pipeline[]>('db_list_pipelines', { companyId });
}

export async function createPipeline(
  companyId: string,
  name: string,
  isDefault = false,
): Promise<Pipeline> {
  return invokeCommand<Pipeline>('db_create_pipeline', {
    companyId,
    name,
    isDefault,
  });
}

// Stages

export async function listDealStages(pipelineId: string): Promise<DealStage[]> {
  return invokeCommand<DealStage[]>('db_list_deal_stages', { pipelineId });
}

export async function createDealStage(input: {
  pipelineId: string;
  name: string;
  position: number;
  probability: number;
  color?: string | null;
}): Promise<DealStage> {
  return invokeCommand<DealStage>('db_create_deal_stage', {
    pipelineId: input.pipelineId,
    name: input.name,
    position: input.position,
    probability: input.probability,
    color: input.color ?? null,
  });
}

// Deals

export async function createDeal(input: CreateDealInput): Promise<Deal> {
  return invokeCommand<Deal>('db_create_deal', {
    companyId: input.companyId,
    contactId: input.contactId ?? null,
    pipelineId: input.pipelineId,
    stageId: input.stageId,
    title: input.title,
    amountMinor: input.amountMinor,
    currency: input.currency ?? null,
    expectedCloseAt: input.expectedCloseAt ?? null,
    notes: input.notes ?? null,
  });
}

export async function getDeal(id: string): Promise<Deal> {
  return invokeCommand<Deal>('db_get_deal', { id });
}

export async function updateDeal(
  id: string,
  companyId: string,
  fields: Partial<Omit<Deal, 'id' | 'company_id' | 'created_at'>>,
): Promise<Deal> {
  // Rust `db_update_deal` requires `company_id` in its DTO to scope the
  // UPDATE to the owning tenant. Omitting it would fail command deserialization.
  return invokeCommand<Deal>('db_update_deal', {
    id,
    companyId,
    fields: fields as unknown as Record<string, unknown>,
  });
}

export async function deleteDeal(id: string): Promise<void> {
  return invokeCommand<void>('db_delete_deal', { id });
}

export async function listDeals(input: {
  companyId: string;
  pipelineId?: string | null;
  stageId?: string | null;
  status?: string | null;
}): Promise<Deal[]> {
  return invokeCommand<Deal[]>('db_list_deals', {
    companyId: input.companyId,
    pipelineId: input.pipelineId ?? null,
    stageId: input.stageId ?? null,
    status: input.status ?? null,
  });
}

export async function moveDealStage(input: MoveDealStageInput): Promise<Deal> {
  return invokeCommand<Deal>('db_move_deal_stage', {
    id: input.id,
    stageId: input.stageId,
  });
}

// Scoring

export async function recomputeScores(input: RecomputeScoresInput): Promise<number> {
  return invokeCommand<number>('db_recompute_scores', {
    companyId: input.companyId,
  });
}

// Idempotently ensure a default pipeline with the standard `DEFAULT_STAGES`
// exists for the company, returning the (existing or newly created) pipeline
// along with its stages. Used by the CRM board on first mount / after reset.
export async function ensureDefaultPipeline(companyId: string): Promise<Pipeline & { stages: DealStage[] }> {
  return invokeCommand<Pipeline & { stages: DealStage[] }>('db_ensure_default_pipeline', {
    companyId,
  });
}

// Fetch a single deal stage by id. Returns `null` if the stage was deleted.
export async function getDealStage(stageId: string): Promise<DealStage | null> {
  return invokeCommand<DealStage | null>('db_get_deal_stage', { stageId });
}
